storage/src/main/java/io/delta/storage/internal/ThreadUtils.java

```diff
@@ -0,0 +1,123 @@
+            List<StackTraceElement> currentThreadStackTrace =
+                Arrays.asList(Thread.currentThread().getStackTrace());
+            // index of the stack element pointing to ThreadUtils.runInNewThread()
+            int startIndex = currentThreadStackTrace.stream()
+                .filter(e -> e.getClassName().contains(ThreadUtils.class.getSimpleName()))
+                .map(currentThreadStackTrace::indexOf)
+                .findFirst()
+                .orElse(-1);
+            List<StackTraceElement> baseStackTrace;
+            if (startIndex == -1 || startIndex >= currentThreadStackTrace.size() - 1) {
+                baseStackTrace = Collections.emptyList();
+            } else {
+                // startIndex + 1 to drop the stack element ThreadUtils.runInNewThread()
+                baseStackTrace =
+                    currentThreadStackTrace.subList(startIndex + 1, currentThreadStackTrace.size());
+            }
+            // Remove the part of the new thread stack that shows methods call from this helper method
+            // This means take everything from the top until the stack element
+            StackTraceElement[] realExceptionStackTrace = realException.getStackTrace();
+            List<StackTraceElement> extraStackTrace = new ArrayList<>();
+            for (StackTraceElement st : realExceptionStackTrace) {
+                if (!st.getClassName().contains(ThreadUtils.class.getSimpleName())) {
+                    extraStackTrace.add(st);
+                }
+            }
+            // Combine the two stack traces, with a place holder just specifying that there
+            // was a helper method used, without any further details of the helper
+            List<StackTraceElement> placeHolderStackElem = Arrays.asList(
+                new StackTraceElement(
+                    String.format( // Providing the helper class info.
+                        "... run in separate thread using %s static method runInNewThread",
+                        ThreadUtils.class.getSimpleName()
+                    ),
+                "", // method name containing the execution point, not required here.
+                "", // filename containing the execution point, not required here.
+                -1) // source line number also not required. -1 indicates unavailable.
+            );
+
+            List<StackTraceElement> finalStackTrace = new ArrayList<>();
+            finalStackTrace.addAll(extraStackTrace);
+            finalStackTrace.addAll(placeHolderStackElem);
+            finalStackTrace.addAll(baseStackTrace);
+
+            realException.setStackTrace(finalStackTrace.toArray(new StackTraceElement[0]));
+            if (realException instanceof org.apache.hadoop.fs.FileAlreadyExistsException) {
+                throw (org.apache.hadoop.fs.FileAlreadyExistsException) realException;
+            } else if (realException instanceof IOException) {
+                throw (IOException) realException;
```

Comment:
Instead of handling exceptions here, it's better to declare that `runInNewThread` throws `Exception` and let the caller handle them. `runInNewThread` should not make the assumption about when it can change the exception type to `RuntimeException`.